### DISCUSSION TITLE
Why does the script precompute private ipv4's instead of having them auto assigned?

### DISCUSSION THREAD

**Discussion by valkenburg-prevue-ch** [mysticaltech/terraform-hcloud-kube-hetzner#541] [Ideas]
Hi,

I see that the script goes to great lengths to precompute the private ips in the various subnets. I've seen this in other projects too. Could this be a relic from the past? One can assign an hcloud_server to a private network and let the private ip be auto assigned. This would remove quite a bit of book keeping, and makes it much easier to add anything to the network.

I have not been able to find bits of code the depend on knowing the private ip before the nodes are created, but perhaps I missed something.

Any thoughts?

**Comment by valkenburg-prevue-ch** [comment]
I checked that indeed auto-assigning works fine, consider this PR: https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/pull/542

I'm open to discussion, should anyone object to the pr.

**Reply by mysticaltech** [reply]
@valkenburg-prevue-ch This is the problem! See https://registry.terraform.io/providers/hetznercloud/hcloud/latest/docs/resources/server_network#argument-reference

![ksnip_20230126-153226](https://user-images.githubusercontent.com/518555/214862689-18176ab7-5c8a-470c-864d-4a10b15497d2.png)

Actually, if the IP is missing, it will add every node to the same subnet, even if another subnet_id is mentioned. In the past, we verified that unfortunate behavior. But they could have fixed this and forgot to update the docs?!

If you deploy a test cluster with your PR, and run `hcloud network describe <cluster-name>` you will actually see the routes that corresponds to each subnet. And then you can verify with `hcloud server describe <node-name>` if each node belong to the right subnet.

In my case the network describe give me two routes, for two nodepools, 1 control-plane, 1 agent nodepool:

![ksnip_20230126-153708](https://user-images.githubusercontent.com/518555/214864128-fe6ce26c-8856-476d-9201-8a12afdd37ce.png)

Please check this on your end!

**Reply by valkenburg-prevue-ch** [reply]
Confirmed, this is very broken... Closed my PR.

**Reply by aleksasiriski** [reply]
> Confirmed, this is very broken... Closed my PR.

Maybe write an email to Hetzner asking for this feature

**Reply by valkenburg-prevue-ch** [reply]
Related question, though: why should control plane and agents be on different subnets? Does that make any difference anywhere?

**Reply by mysticaltech** [reply]
If I remember well, we tried mixing it up at the beginning, but if I remember well, there was also an issue when it could give the same IP when two servers were created at the same time. Now, maybe that has been fixed, as I do not see mentions of it anymore! Honesly, that is a vague memory.

But the way it is now is also clear, as each nodepool corresponds to a subnet. It's more organized, and I guess makes the underlying routing easier for the CNI. Just suppositions.

But definitely, if they can give the IP dynamically in the right subnet, it would be great to move to that.

**Reply by mysticaltech** [reply]
Please @valkenburg-prevue-ch, feel free to open a ticket to Hetzner!

**Reply by ifeulner** [reply]
It makes sense to have different subnets at least for control nodes and agents, in a production installation you would have a firewall between these node types.
But IP handling in general is an open issue, as we should also allow configuration options for pods, services, etc. 
Besides that I would prefer to have as much automatic assigning as possible...

**Reply by mysticaltech** [reply]
@ifeulner Definitely, automatic is better, but as explained [in this comment](https://github.com/orgs/kube-hetzner/discussions/541#discussioncomment-4788260), we did not have a choice here.

**Comment by valkenburg-prevue-ch** [comment]
It's been a while, but I finally got around contacting support and now reporting back here. In short: auto-assigning IPs to different pre-created subnets is not going to be supported soon, but it is on Hetzner's radar as per this new issue: https://github.com/hetznercloud/terraform-provider-hcloud/issues/672 .


> 
> Dear customer,
> 
> this is not easily possible with our current API, as the relevant endpoint only allows assigning a server to a network, not to a specific subnet. Passing an IP "bypasses" this, but someone needs to calculate that IP, right now this responsibility is passed to you as the users of the terraform module.
> 
> I talked to the responsible team for this endpoint and they stated that they have no plans to support assigning a server into a specific subnet at this time.
> 
> I have opened an issue on our public GitHub issue tracker that explains the problem, and we are open to pull requests that implement this.
> 
> https://github.com/hetznercloud/terraform-provider-hcloud/issues/672
> 
> 
> Mit freundlichen Grüßen / Kind regards
> 
> Julian  
> 
> > --------------------------------------------------
> > Dear madam, sir,
> >
> > as per
> > https://registry.terraform.io/providers/hetznercloud/hcloud/latest/docs/resources/server_network#argument-reference
> > , you state  "subnet_id - (Optional, string) ID of the sub-network which
> > should be added to the Server. Required if network_id is not set. Note: if
> > the ip property is missing, the Server is currently added to the last
> > created subnet."
> >
> > Is there any chance this issue in the note can be repaired? That is, auto
> > ip assignment on any chosen pre-existing subnet, rather then overriding the
> > subnet_id to be the last created subnet? Our project currently pre-computes
> > all ips of the created instances, but this is quite some code to maintain
> > and makes it cumbersome to add new features.
> >
> > Many thanks in advance!
> >
> > Kind regards,
> > Wessel 
>

**Reply by mysticaltech** [reply]
@valkenburg-prevue-ch Wonderful, well done! Apricot (the handle for Julian I guess) is pretty active. This ball has a good change of rolling 🤞🤞
